### PR TITLE
Add Edit User Form

### PR DESCRIPTION
**Pull request by conradbeach**
## Goal
Pivotal Tracker: [#171524491](https://www.pivotaltracker.com/story/show/171524491), [#171525027](https://www.pivotaltracker.com/story/show/171525027)

This adds a modal and form that allows an admin to update the following information for a user:
- Name
- Login ID
- Password 
- SIS ID
- Email

## Implementation
For a couple of the API calls to Canvas, I used `canvas_api.api_put_request` instead of `canvas_api.proxy`. I generally prefer using `proxy`, but it wasn't working for those 2 API calls for whatever reason. I'd like to figure out what the issues were, but I didn't want to let it hold up adding this code.

## Things to Look Out for While Reviewing
When we update the name, the sortable name is updated automatically, but the display name (AKA short name) is not. Should we update the display name too?

With this implementation, when we update the student email, it actually just adds a new email to their account and makes it their primary email. It doesn't remove their old one. I think this is fine.
It also doesn't confirm the email. This also makes sense to me.
Does anyone have objections to how this implementation handles emails?

The `update` action issues 3 separate requests to the Canvas API. Not terrible, but not great. If anyone knows of a way to reduce this number, I'd appreciate suggestions.

## Demo
update_student

(This demo doesn't show updating the password because I didn't see an easy way to actually show that working; I did test changing a password though and verified that it does work.)

### CODE REVIEW

File: client/apps/user_tool/reducers/application.js [hunk 1]
```diff
@@ -19,7 +19,38 @@ export default (state = initialState(), action) => {
...
+        nextPageAvailable
+      };
+    }
+
+    case ApplicationConstants.UPDATE_USER_DONE: {
+      const {
+        id,
+        name,
+        login_id,
+        sis_user_id,
```
[the remark this answers is not in the document]
Comment (conradbeach) [reply client/apps/user_tool/reducers/application.js]: I don't think it makes sense to change it to camel case and then change it right back to snake case.

File: client/apps/user_tool/reducers/application.js [hunk 2]
```diff
@@ -19,7 +19,38 @@ export default (state = initialState(), action) => {
...
+        previousPageAvailable,
+        nextPageAvailable
+      };
+    }
+
+    case ApplicationConstants.UPDATE_USER_DONE: {
+      const {
+        id,
+        name,
+        login_id,
```
[the remark this answers is not in the document]
Comment (conradbeach) [reply client/apps/user_tool/reducers/application.js]: I don't think it makes sense to change it to camel case and then change it right back to snake case.

File: client/apps/user_tool/actions/application.js [hunk 3]
```diff
@@ -18,3 +18,26 @@ export const searchForAccountUsers = (lmsAccountId, searchTerm, page) => ({
...
+      name: userAttributes.name,
+      login_id: userAttributes.loginId,
+      sis_user_id: userAttributes.sisUserId,
+      email: userAttributes.email,
+    }
+  };
+
+  if (userAttributes.password !== '') {
+    body.user.password = userAttributes.password;
+  }
```
Comment (conradbeach) [inline client/apps/user_tool/actions/application.js]: Conditionally adding the `password` property like this isn't strictly necessary; it's just me being cautious. If we pass an empty string for the password to the Canvas API, it just ignores it and doesn't change the user's password.

**Comment by conradbeach** [comment]
CI isn't running again, but I ran tests locally and they all pass.

**Comment by bfcoder** [comment]
There is a failing javascript spec https://travis-ci.org/github/atomicjolt/adhesion a syntax error.

**Comment by conradbeach** [comment]
@bfcoder The build you linked to is actually a different branch (`ckb/add-no-results-components`).

The build for the branch associated with the PR is green.
image

**Comment by bfcoder** [comment]
ahh, ok

**Comment by conradbeach** [comment]
I didn't think to just go check the build on Travis. I assumed that it wasn't running at all if it wasn't reporting here, but I was obviously wrong about that. I'll check Travis in the future if it doesn't report on the PR page.

**Comment by bfcoder** [comment]
We need to authorize the user making edits to a canvas user. Make sure they are an admin on the update action for the canvas users controller.

**Comment by bfcoder** [comment]
In other words we still need to do server side authorization. So someone can't craft a PATCH to that endpoint and make unauthorized changes.

File: app/controllers/api/canvas_users_controller.rb [hunk 4]
```diff
@@ -22,4 +22,54 @@ def index
       status: :ok,
     )
   end
+
+  def update
+    edit_user_response = edit_user_on_canvas
+
+    # This ID is the ID of the login record; it's different from the user's login ID.
+    numeric_login_id = find_canvas_user_login["id"]
```
[the remark this answers is not in the document]
Comment (bfcoder) [reply app/controllers/api/canvas_users_controller.rb]: Well I can't seem to figure out how to make a person not have any logins. So it probably isn't an issue.

[the remark this answers is not in the document]
Comment (conradbeach) [reply app/controllers/api/canvas_users_controller.rb]: Ok

**Comment by conradbeach** [comment]
@bfcoder I'll take a closer look at the authorization.

**Comment by conradbeach** [comment]
@bfcoder I don't have a great understanding of how authentication works inside our applications. Thus far, I've been developing the user tool with a `CANVAS_TOKEN` set in my `.env`. My understanding is that the application is using that token to make the requests to the Canvas API.

My plan was to, later, remove that token and do more work to verify the permissions of the current user. Could I remove that token from the environment and instead depend on getting a token from Canvas when the user authorizes the application inside of Canvas? I've observed other applications of ours performing this type of authorization though I don't know what exactly it does. In my head, I'm imagining that we could use the token from the user and it would already be scoped to whatever permissions they have and we wouldn't have to perform an additional check in the controller.

My understanding is likely incorrect or incomplete in some way on this though.

Also, could you point me to an example implementation of the type of server side authorization you're describing?

**Comment by bfcoder** [comment]
You will use `current_user.lti_admin?(context_id)` to authorize them.

**Comment by conradbeach** [comment]
Ok. Thanks.

**Comment by conradbeach** [comment]
@bfcoder I added `current_user.lti_admin?(context_id)` to a before action that returns a `401 Unauthorized` if the user is not an admin. This prevents non-admins from accessing user data. However, it doesn't prevent admins from accessing user data from accounts they're not admins in. After logging in as a sub-account admin, I was able to modify the request and have the application list and update users from the root account even though I don't have access to that account.

Do you have suggestions on how to prevent this type of abuse?

**Comment by bfcoder** [comment]
hmm, the context_id is suppose to make sure they can only access the account for which they are an account admin.

File: app/controllers/api/canvas_users_controller.rb [hunk 5]
```diff
@@ -22,4 +17,83 @@ def index
...
+    end
+  end
+
+  def search_for_users_on_canvas(search_term, page = nil)
+    # We're manually constructing the URL here and using `api_get_request` instead of
+    # `canvas_api.proxy("LIST_USERS_IN_ACCOUNT", params)` because `proxy("LIST_USERS_IN_ACCOUNT")`
+    # doesn't support the `include` param since it's undocumented.
+    canvas_url = "accounts/#{params[:canvas_account_id]}/users" \
+      "?search_term=#{search_term}&include[]=email"
+    canvas_url += "&page=#{page}" if page # You get a 404 if you pass `&page=`
```
Comment (bfcoder) [inline app/controllers/api/canvas_users_controller.rb:67]: Consider using `to_query`?
```suggestion
    query_params = {
      search_term: search_term,
      include: [:email],
    }
    query_params[:page] = page if page.present? # You get a 404 if you pass `&page=`
    account_id = params[:canvas_account_id]
    canvas_url = "accounts/#{account_id}/users?#{query_params.to_query}"
```

Comment (conradbeach) [reply app/controllers/api/canvas_users_controller.rb:67]: Yeah, that's a nice improvement. Thanks. I'll add it.

Comment (conradbeach) [reply app/controllers/api/canvas_users_controller.rb]: Added

**Comment by conradbeach** [comment]
Thanks!

**Pull request merged**